src/main/scala/me/rexim/morganey/ReplAutocompletion.scala

```diff
@@ -0,0 +1,153 @@
+
+    def stripExtensionIfModuleFile(f: File): File =
+      if (isMorganeyModule(f)) new File(f.getParent, f.getName.replaceAll(s".$fileExtension", ""))
+      else f
+
+    def everythingIn(path: List[String], fileNameFilter: String => Boolean = _ => true) =
+      findAllModulesIn(path.mkString("."))
+        .map { case (root, f) => (root, stripExtensionIfModuleFile(f)) }
+        .filter { case (root, f) => fileNameFilter(f.getName) }
+        .map(relativize)
+        .map(relativeFileToLoadPath)
+
+    def relativize(baseAndFile: (File, File)): String = {
+      val (base, file) = baseAndFile
+      base.toURI().relativize(file.toURI()).getPath()
+    }
+
+    (parts, endsWithDot) match {
+      // load .|
+      case (Nil, true)       => Nil
+      // load a.b.|
+      case (xs, true)        => everythingIn(xs)
+      // load |
+      case (Nil, false)      => topLevelDefinitions().map(stripExtensionIfModuleFile).map(_.getName)
+      // load math.ari|
+      case (xs :+ x, false)  => everythingIn(xs, s => s.toLowerCase startsWith x.toLowerCase)
+    }
+  }
+
+  private object LoadStatement {
+    private val loadStatement = "load (.*)".r
+
+    def unapply(line: String): Option[(List[String], Boolean)] =
+      Option(line) collect {
+        case loadStatement(modulePath) =>
+          (modulePath.split("\\.").toList, modulePath.endsWith("."))
+      }
+  }
+
+  private def matchingDefinitions(line: String, knownVariableNames: List[String]): List[String] =
+    for {
+      name <- knownVariableNames
+      matchLength = countPrefixMatch(line, name)
+      if matchLength > 0
+      prefix = line dropRight matchLength
+    } yield prefix + name
+
+  private def countPrefixMatch(line: String, definition: String): Int = {
+    @tailrec
+    def countMatchingLetters(xs: List[Char], ys: List[Char], acc: Int): Int = (xs, ys) match {
+      case (xh :: xt, yh :: yt) if xh == yh => countMatchingLetters(xt, yt, acc + 1)
+      case _                                => acc
+    }
+
+    def normalize(s: String) = s.toLowerCase.toList
+    val lineLowerCase = line.toLowerCase
+
+    /**
+      * Only autocomplete:
+      * - If the user typed in a single incomplete name: autocomplete if the name,
+      *    which was typed into the repl is a prefix of the name of a definition
+      * - If the user typed in a single name, which is part of a complex term
+      */
+    def keepNamesAndComplexTerms(nameInLine: String): Boolean = nameInLine match {
+      // input of repl is exactly `nameInLine`
+      case s if s == lineLowerCase     => definition startsWith lineLowerCase
+      // REPL has a complex term as input
+      case s if s.length < line.length => true
+      case _                           => false
+    }
+
+    lastNameInLine(line)
+      .filter(keepNamesAndComplexTerms)
+      .map(ln => countMatchingLetters(normalize(definition), normalize(ln), 0))
+      .getOrElse(0)
+  }
+
+  private val namePattern = "[a-zA-Z][a-zA-Z0-9]*"
```

Comment:
I think that calling `.r` here will save a bit of performance (because regex will be created ahead and cached between calls).